### ISSUE TITLE
rustdoc stack overflow on mutually recusive Deref implementation

### ISSUE THREAD

**Issue by trinity-1686a** [rust-lang/rust#85095]
I tried this code:

```rust
use std::ops::Deref;

pub struct A;
pub struct B;

impl Deref for A {
    type Target = B;

    fn deref(&self) -> &Self::Target {
        panic!()
    }
}

impl Deref for B {
    type Target = A;

    fn deref(&self) -> &Self::Target {
        panic!()
    }
}
```
Trying to run rustdoc with `rustdoc lib.rs`

I expected to see this happen: I'm not sure actually, probably an ICE, but I'd prefer it either detect and accept recursion, or detect and reject with a proper error message.

Instead, this happened: I get a stack overflow
```
thread 'rustc' has overflowed its stack
fatal runtime error: stack overflow
```
GDB tells me `rustdoc::html::render::sidebar_deref_methods` is on the stack around 17 thousand times
(when running in GDB, it actually segfaulted at `<alloc::vec::Vec<T> as alloc::vec::spec_from_iter::SpecFromIter<T,I>>::from_iter` instead of stack-overflowing)

`impl Deref<Target=A> for A` does not seems to pose problems

### Meta

`rustc --version --verbose`:
```
rustc 1.52.0 (88f19c6da 2021-05-03)
binary: rustc
commit-hash: 88f19c6dab716c6281af7602e30f413e809c5974
commit-date: 2021-05-03
host: x86_64-unknown-linux-gnu
release: 1.52.0
LLVM version: 12.0.0
```

**Comment by jyn514** [comment]
This should be "fixed" by https://github.com/rust-lang/rust/pull/84867. cc https://github.com/rust-lang/rust/pull/80653

**Comment by trinity-1686a** [comment]
I tested by cloning the pr, running `./x.py build --stage 1` and executing `./build/x86_64-unknown-linux-gnu/stage1/bin/rustdoc lib.rs` and the issue is still here. Either I'm missing something, or if fixes only a close, but different issue

**Comment by trinity-1686a** [comment]
I've done more testing, and #84867 does solve part of the issue.
When using that branch, the infinite recursion happen in the exact same place, nothing changed. I've made a small patch (link bellow) that, when applied on master, solve this stack overflow, but another one happen on `rustdoc::passes::collect_trait_impls::collect_trait_impls::add_deref_target`. Trying the same patch on #84867 and the issue is really fixed, so #84867 fix one stack overflow, but there is an other that is triggered before.


[0001-fix-infinite-recursion-on-sidebar_deref_methods.patch.gz](https://github.com/rust-lang/rust/files/6447606/0001-fix-infinite-recursion-on-sidebar_deref_methods.patch.gz) (gziped as Github wouldn't accept it otherwise)
